atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraWriteTestSuite.java

```diff
@@ -0,0 +1,147 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.ete;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import org.assertj.core.util.Lists;
+import org.junit.After;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.palantir.atlasdb.containers.CassandraEnvironment;
+import com.palantir.atlasdb.http.errors.AtlasDbRemoteException;
+import com.palantir.atlasdb.todo.ImmutableTodo;
+import com.palantir.atlasdb.todo.Todo;
+import com.palantir.atlasdb.todo.TodoResource;
+
+public class QosCassandraWriteTestSuite extends EteSetup {
+    private static final List<String> CLIENTS = ImmutableList.of("ete1");
+
+    @ClassRule
+    public static final RuleChain COMPOSITION_SETUP = EteSetup.setupComposition(
+            QosCassandraWriteTestSuite.class,
+            "docker-compose.qos.cassandra.yml",
+            CLIENTS,
+            CassandraEnvironment.get());
+
+    @Test
+    public void shouldFailIfWritingTooManyBytes() {
+        TodoResource todoClient = EteSetup.createClientToSingleNode(TodoResource.class);
+
+        ensureOneWriteHasOccurred(todoClient);
+
+        assertThatThrownBy(() -> todoClient.addTodo(getTodoOfSize(100_000)))
+                .isInstanceOf(RuntimeException.class)
+                .as("Cannot write 100_000 bytes the second time as write limit of "
+                        + "1000 was consumed and the burst isnt enough either.");
+    }
+
+    private void ensureOneWriteHasOccurred(TodoResource todoClient) {
+        try {
+            todoClient.addTodo(getTodoOfSize(100_000));
+            // okay as the first huge write is not rate limited.
+        } catch (Exception e) {
+            // okay as some other test might have written before
+        }
+    }
+
+    @Test
+    public void canNotWriteLargeNumberOfBytesConcurrentlyIfAllRequestsComeAtTheExactSameTime()
+            throws InterruptedException {
+        TodoResource todoClient = EteSetup.createClientToSingleNode(TodoResource.class);
+
+        CyclicBarrier barrier = new CyclicBarrier(100);
+        ForkJoinPool threadPool = new ForkJoinPool(100);
+        List<Future<?>> futures = Lists.newArrayList();
+
+        IntStream.range(0, 100).parallel().forEach(i ->
```

Comment:
dont need parallel here

Reply:
there's a 👍 , yet we still have parallel...

Reply:
The test was removed.